File: backend/app/services/llm/gemini_provider.py
import json
import asyncio
from typing import Type, TypeVar
from pydantic import BaseModel
from .provider import LLMProvider

T = TypeVar("T", bound=BaseModel)


class GeminiProvider(LLMProvider):
    name = "gemini"

    def __init__(self, api_key: str):
        import google.generativeai as genai
        genai.configure(api_key=api_key)
        self.genai = genai
        self.model = genai.GenerativeModel("gemini-2.0-flash")

    async def generate_structured(
        self,
        prompt: str,
        system_prompt: str,
        schema: Type[T]
    ) -> T:
        full_prompt = f"{system_prompt}\n\n{prompt}\n\nReturn JSON ONLY matching the requested schema."
        
        def _generate():
            return self.model.generate_content(
                full_prompt,
                generation_config={"response_mime_type": "application/json"}
            )

        response = await asyncio.to_thread(_generate)
        text = response.text
        data = json.loads(text)
        return schema.model_validate(data)

File: backend/src/services/llm/geminiProvider.ts
import { z } from 'zod';
import { LLMProvider } from './provider.js';

export class GeminiProvider implements LLMProvider {
  name = 'gemini';
  private model: any;

  constructor(apiKey: string) {
    try {
      const { GoogleGenerativeAI } = require('@google/generative-ai');
      const genAI = new GoogleGenerativeAI(apiKey);
      this.model = genAI.getGenerativeModel({ model: 'gemini-1.5-flash' });
    } catch (err) {
      console.warn('Gemini SDK initialization warning:', err);
    }
  }

  async generateStructured<T>(prompt: string, systemPrompt: string, schema: z.ZodType<T>): Promise<T> {
    if (!this.model) {
      throw new Error('GeminiProvider not initialized');
    }

    const fullPrompt = `${systemPrompt}\n\n${prompt}\n\nReturn JSON ONLY matching the requested schema. Do not include raw markdown block wrappers.`;
    const result = await this.model.generateContent({
      contents: [{ role: 'user', parts: [{ text: fullPrompt }] }],
      generationConfig: { responseMimeType: 'application/json' },
    });

    const rawText = result.response.text();
    const cleanText = rawText.replace(/```json/gi, '').replace(/```/g, '').trim();
    const data = JSON.parse(cleanText);
    return schema.parse(data);
  }
}
